extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/configproperties/ClassConfigPropertiesUtil.java

```diff
@@ -0,0 +1,359 @@
+
+    private ClassConfigPropertiesUtil() {
+    }
+
+    /**
+     * Generates a class like the following:
+     *
+     * <pre>
+     * &#64;ApplicationScoped
+     * public class EnsureValidation {
+     *
+     *     &#64;Inject
+     *     MyConfig myConfig;
+     *
+     *     &#64;Inject
+     *     OtherProperties other;
+     *
+     *     public void onStartup(@Observes StartupEvent ev) {
+     *
+     *     }
+     * }
+     * </pre>
+     *
+     * This class is useful in order to ensure that validation errors will prevent application startup
+     */
+    static void generateStartupObserverThatInjectsConfigClass(ClassOutput classOutput, Set<DotName> configClasses) {
+        try (ClassCreator classCreator = ClassCreator.builder().classOutput(classOutput)
+                .className(ConfigPropertiesUtil.PACKAGE_TO_PLACE_GENERATED_CLASSES + ".ConfigPropertiesObserver")
+                .build()) {
+            classCreator.addAnnotation(ApplicationScoped.class);
+
+            for (DotName configClass : configClasses) {
+                String configClassStr = configClass.toString();
+                FieldCreator fieldCreator = classCreator
+                        .getFieldCreator(
+                                configClass.isInner() ? configClass.local()
+                                        : configClass.withoutPackagePrefix() + "_" + HashUtil.sha1(configClassStr),
+                                configClassStr)
+                        .setModifiers(Modifier.PUBLIC); // done to prevent warning during the build
+                fieldCreator.addAnnotation(Inject.class);
+            }
+
+            try (MethodCreator methodCreator = classCreator.getMethodCreator("onStartup", void.class, StartupEvent.class)) {
+                methodCreator.getParameterAnnotations(0).addAnnotation(Observes.class);
+                methodCreator.returnValue(null); // the method doesn't need to do anything
+            }
+        }
+    }
+
+    /**
+     * @return true if the configuration class needs validation
+     */
+    static boolean addProducerMethodForClassConfigProperties(ClassInfo configPropertiesClassInfo,
+            ClassCreator producerClassCreator, String prefixStr, IndexView applicationIndex) {
+
+        if (!DotNames.OBJECT.equals(configPropertiesClassInfo.superName())) {
+            throw new IllegalArgumentException(
+                    "Classes annotated with @" + DotNames.CONFIG_PROPERTIES
+                            + " cannot extend other classes. Offending class is "
+                            + configPropertiesClassInfo);
+        }
+
+        if (!configPropertiesClassInfo.hasNoArgsConstructor()) {
+            throw new IllegalArgumentException(
+                    "Class " + configPropertiesClassInfo + " which is annotated with " + DotNames.CONFIG_PROPERTIES
+                            + " must contain a no-arg constructor");
```

Comment:
What? You don't generate the no-args constructor automatically? Are you Georgios or what? :)

Reply:
Lololololo!

I actually hadn't thought of what, but I don't think it's possible because we would miss out on field initialization which in this case matters